### ISSUE TITLE
f-strings for fmt in C++

### ISSUE THREAD

**Issue by hadrielk** [fmtlib/fmt#3691]
This is going to be a very strange `fmt` issue...

I've been working on a standards proposal for WG21 to add f-string support for C++. I.e., similar to python f-strings but for `std::format()` - though it can work with `fmt::format()` as well, or `fmt::print()`, etc. It's basically just string-interpolation for C++, for the format syntax.

I wasn't sure if there was interest in the idea, so I submitted a straw-man draft on the [std-proposals mailing list](https://lists.isocpp.org/mailman/listinfo.cgi/std-proposals), and there was interest.

Based on email exchanges there, I've updated the draft a couple times. I'm currently finishing up the [third version of the draft](http://api.csswg.org/bikeshed/?url=https://raw.githubusercontent.com/hadrielk/cpp-proposals/main/f-string/f-string-r2.bs), though it's still a work-in-progress.

Given that @vitaut was the main driver behind all this getting into C++ in the first place, and `fmt` was the basis for `std::format`, I'm posting this issue in the hopes of getting feedback for whether this is a good idea or not.

---

### The TL;DR

Before this proposal:
```c++
std::string stringify(std::string_view prefix, int bits) {
    return std::format("{}:{}: got {} for {:#06x}", prefix, __LINE__, calculate(bits), bits);
}

void display(std::string_view prefix, int bits) {
    std::print("{}:{}: got {} for {:#06x}", prefix, __LINE__, calculate(bits), bits);
}
```

After this proposal:
```cpp
std::string stringify(std::string_view prefix, int bits) {
    return F"{prefix}-{__LINE__}: got {calculate(bits)} for {bits:#06x}";
}

void display(std::string_view prefix, int bits) {
    std::print(X"{prefix}-{__LINE__}: got {calculate(bits)} for {bits:#06x}");
}
```

**Comment by vitaut** [comment]
Thanks for working on this! There has been some interest in string interpolation and the format string syntax has been chosen with named arguments and interpolation in mind.

I haven't looked into the proposal in detail but here's a few quick comments/questions:

> arguably more popular than the python `str.format()`

You probably need to give some data to back this claim.

Why are `F` and `X` prefixes capitalized?

I would recommend looking at interpolation in Rust (https://rust-lang.github.io/rfcs/2795-format-args-implicit-identifiers.html) especially since their formatting facility is closer to what we have in C++. Python is somewhat limited because it only works with strings. One particularly interesting feature of Rust's facility is that it completely avoids any prefixes which is much more convenient.

**Comment by hadrielk** [comment]
> You probably need to give some data to back this claim.

A search on grep.app shows f-strings have a lead over format, but yeah it's only ~15% more hits. So I think I'll just remove that statement in the next rev. I was only trying to explain that they're very popular, anyway, for those in C++ who might not know.

> Why are F and X prefixes capitalized?

Only to match the capitalization of `R` for raw-strings, and `L`/`U` encodings. I assumed the WG would ultimately decide what letter(s) to use?

> One particularly interesting feature of Rust's facility is that it completely avoids any prefixes which is much more convenient.

The challenge with this for C++, is that if the extraction is delayed until phases-7/8 of translation, then macros cannot be used within the original interpolation-string.

I.e., if we just used normal string-literals, and then left it up to `std::format_string<>` to handle - or some intermediate C++ thing that created a `std::format_string<>` and args - then that's already past the point of preprocessing, so all macros no longer exist.

Which means things like `__FILE__` and `__LINE__` and such cannot be used within the braces replacement-field, nor can we support expressions without causing confusion - because some functions are function-like macros not real C++ functions, and programmers often don't even know which type they are. (for example the old `abs()` and `min()`/`max()` are sometimes function-like macros, for some compiler cstdlibs)

So for example, this would not work:
```cpp
std::print("{prefix}-{__LINE__}: got {calculate(bits)} for {bits:#06x}");
```

---

The other main advantage to handling this in the preprocessor vs. in later phases, is that it could potentially be used in C. And I've been told that's not as unlikely as it may seem. (of course it wouldn't be usable in the C++ functions, nor would it have the implicit-formatting of arguments based on their types, but there are other ways it can be used in C)

---

As an aside: if we did it the "interpolation" in the later compilation phases, it would impact the runtime performance a bit. Because the string-literal put into static storage would be the original pre-interpolation one, not the post-interpolation format-string. So that would need to be re-parsed each time it's used. Of course that happens today too, but the syntax introduced by this proposal would be more expensive to parse, especially if we have to handle quoted-strings within it. (though that's still TBD anyway)

Whereas dong it in the preprocessor means the string-literal that's put into static storage and used by runtime functions, is the post-interpolation format-string of today.

**Comment by rmu75** [comment]
How would this work with internationalization i.e. how would it be possible to translate the format strings?

**Comment by hadrielk** [comment]
> How would this work with internationalization i.e. how would it be possible to translate the format strings?

Good question. I've been asked about that and we don't have an answer yet. We're tracking that issue in [string-interpolation issue #2](https://github.com/hadrielk/string-interpolation/issues/2).

So far we've been asked to make it work with POSIX `gettext()`, which I think is possible. 

But making it work with the non-posix gnu-lib extension `ngettext()` (note the '`n`') is cumbersome and might not be practical.

Are those the types of things you're thinking of when you say "work with internationalization"? Or are you thinking of something else?

**Comment by rmu75** [comment]
Yes I'm thinking of some form of gettext.

**Comment by DanielaE** [comment]
I've been working on gettext-based internationalization for std::format for quite some time now. The user-facing frontend is there and I've replaced the Boost.Locale frontend in our codebase with mine. It's now in production on all customer sites.
I can do

- compile-time argument checking like format/print
- runtime translation, consistency checking of format-specs in translated strings is done by the translation toolchain
- compile-time argument checking for the plural format string
- plurals without a '`n`' argument to the format string, also checked at compile-time for consistency and type-safety
- contexts
- formatting: `char` with mandatory UTF-8, `wchar_t` with mandatory UTF-16
- `.po` with mandatory UTF-8
- no encoding conversions in the formatting process

**Comment by hadrielk** [comment]
> I've been working on gettext-based internationalization for std::format for quite some time now.

Great! Do you mind helping us with figuring out how to handle these use-cases?

In particular things like handling plural forms. And also index-numbered arguments needs for internationalization.

We've started an issue for this topic here: [International translation issues](https://github.com/hadrielk/string-interpolation/issues/2).

Any help you can provide would be much appreciated!

**Comment by DanielaE** [comment]
I see very little meat there, right?
What kind of formatting syntax do you envision? The C++20 format-spec syntax, or something else? The former has support in xgettext and e.g. POEdit. The latter would be a hard sell without tooling.
It seems like you want to build something around preprocessor macros. This is a dead end for me because that is inherently incompatible with modules.

You might consider looking [here](https://hachyderm.io/deck/@DanielaKEngert/110774612852584146) to see where I'm at right now.

**Comment by DanielaE** [comment]
> Out of curiosity, how did you get `println()` to accept a runtime string?

This is not a runtime string. It is a `basic_format_string_translator` that is constructed at compile time from `translate` and is convertible to a `std::basic_format_string`.

All the magic happens in `basic_format_string_translator` at both compile time and runtime.

It looks like we have different interests. You're after f-strings which is of mild interest to me, but solid support for translation is paramount to me.

**Comment by vitaut** [comment]
I would recommend not introducing yet another prefix for translation. You should be able to achieve what Daniela described with just `X` and a separate API that better not be in your paper either because it is unrelated to interpolation.

**Comment by vitaut** [comment]
Closing the issue since it's not actionable on the {fmt} side and will send some comments via email.
